src/noisepy/seis/datatypes.py

```diff
@@ -0,0 +1,161 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import List
+
+import numpy as np
+from datetimerange import DateTimeRange
+
+from noisepy.seis import noise_module
+
+
+@dataclass
+class ChannelType:
+    """
+    A type of channel, e.g. BHN, but not associated with a particular station
+    """
+
+    name: str
+
+    def __init__(self, name):
+        assert len(name) == 3, "A channel type name should be length 3"
+        self.name = name
+
+    def get_orientation(self) -> str:
+        return self.name[-1]
+
+
+@dataclass
+class Station:
+    network: str  # 2 chars
+    name: str  # 3-5 chars
+    lat: float
+    lon: float
+    elevation: float
+    location: str
+
+    def __repr__(self) -> str:
+        return f"{self.network}.{self.name}"
+
+
+class CorrelationMethod(Enum):
+    XCORR = 1
+    DECONV = 2
+
+
+@dataclass
+class CrossCorrelationParameters:
+    method: CorrelationMethod
+    substack: bool
+    # other parameters here
+
+
+@dataclass
+class Channel:
+    """
+    A channel instance belonging to a station. E.g. CI.ARV.BHN
+    """
+
+    type: ChannelType
+    station: Station
+
+    def __repr__(self) -> str:
+        return f"{self.station}.{self.name}"
+
+
+class DataSource(ABC):
+    """
+    A base abstraction over a data source for seismic data
+    """
+
+    @abstractmethod
+    def get_stations() -> List[Station]:
+        pass
+
+    @abstractmethod
+    def get_channel_types(self) -> List[ChannelType]:
+        pass
+
+    @abstractmethod
+    def get_channels(self) -> List[Channel]:
+        pass
+
+    @abstractmethod
+    def get_timespans(self) -> List[DateTimeRange]:
+        pass
+
+
+class RawDataSource(DataSource):
+    """
+    A class for reading the raw data for a given channel.
+    TODO: write support
+    """
+
+    @abstractmethod
+    def read_data(self, timespan: DateTimeRange, chan: Channel) -> np.ndarray:
+        pass
+
+
+class CrossCorrelationDataStore(DataSource):
+    @abstractmethod
+    def contains(self, timespan: DateTimeRange, chan1: Channel, chan2: Channel):
+        pass
+
+    @abstractmethod
+    def append(
+        self,
+        timespan: DateTimeRange,
+        chan1: Channel,
+        chan2: Channel,
+        parameters: CrossCorrelationParameters,
+        data: np.ndarray,
+    ):
+        pass
+
+    @abstractmethod
+    def read(self, chan1: Channel, chan2: Channel, start: datetime, end: datetime) -> np.ndarray:
+        pass
+
+
```

Comment:
Below is a pseudo-implementation to illustrate how the core code might change as a result